api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java

```diff
@@ -0,0 +1,211 @@
+		log.info("Executing rule to find Patients with similar TB identifiers: duplicated across multiple patients");
+		String queryString = "SELECT o FROM Obs o WHERE o.voided = false AND o.encounter.patient.dead = 0 AND o.encounter.voided = 0 AND o.concept.uuid = :identifierConceptUuid AND o.valueText IN " 
+				+ " (SELECT ob.valueText FROM Obs ob WHERE ob.concept.uuid = :identifierConceptUuid AND ob.voided = 0 GROUP BY ob.concept.conceptId, ob.valueText HAVING COUNT(ob.valueText) > 1)"; 
+		Query query = getSession().createQuery(queryString);
+		query.setParameter("identifierConceptUuid", identifierConceptUuid);
+		
+		List<Obs> obsList = query.list();
+		Set<Patient> uniquePatientList = new HashSet<Patient>();
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Obs obs : obsList) {
+			Patient patient = obs.getEncounter().getPatient();
+			
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
+			ruleResult.setNotes("The " + identifierTitle + "[" + getTbNumber(patient, obs.getEncounter()) + "] is used by another patient");
+			ruleResult.setEntity(patient);
+			
+			ruleResults.add(ruleResult);
+			uniquePatientList.add(patient);
+		}
+		log.info("There are " + uniquePatientList.size() + " Patients with similar " + identifierTitle);
+		
+		return ruleResults;
+	}
+
+	/**
+	 * Patients with similar TB identifiers: duplicated for a single patient,
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> singlePatientWithDuplicateTBNumberAcrossMultipleTreatmentPrograms(String identifierConceptUuid, String identifierTitle) {
+		log.info("Executing rule to find Patients with similar TB identifiers: duplicated for a single patient");
+		String queryString = "SELECT o FROM Obs o WHERE o.voided = false AND o.encounter.patient.dead = 0 AND o.encounter.voided = 0 AND o.concept.uuid = :identifierConceptUuid AND o.obsId IN " 
+				+ " (SELECT min(ob.obsId) FROM Obs ob WHERE ob.concept.uuid = :identifierConceptUuid AND ob.voided = 0 GROUP BY ob.concept.id, ob.valueText HAVING COUNT(ob.valueText) > 1)"; 
+		
+		Query query = getSession().createQuery(queryString);
+		query.setParameter("identifierConceptUuid", identifierConceptUuid);
+		
+		List<Obs> obsList = query.list();
+		Set<Patient> uniquePatientList = new HashSet<Patient>();
+		
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Obs obs : obsList) {
+			Patient patient = obs.getEncounter().getPatient();
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + obs.getEncounter().getId());
+			ruleResult.setNotes("The " + identifierTitle + "[" + getTbNumber(patient, obs.getEncounter()) + "] is used by another patient");
```

Comment:
@moshonk 

1. The unit TB number is returned irrespective of what identifier is selected see attachment 

2. The Patient # (OpenMRS ID) not being completed for no identifiers entered 

3. Remove the [] in which the TB identifiers are displayed 
<img width="524" alt="screenshot 2017-08-05 19 13 05" src="https://user-images.githubusercontent.com/689900/28996938-7ad84e2a-7a12-11e7-8278-5a1c013672b9.png">